converters/toZigbee.js

```diff
@@ -1429,6 +1429,80 @@ const converters = {
             await entity.read('closuresDoorLock', ['lockState']);
         },
     },
+    pincode_lock: {
+        key: ['pin_code'],
+        convertSet: async (entity, key, value, meta) => {
+            const userId = value.userId;
+            const pinCode = value.pinCode;
+            if ( isNaN(userId) ) {
+                throw new Error('userId must be numbers');
+            }
+            if (converters.pincode_lock.userIdInRange(meta, userId)) {
+                throw new Error('userId must be in range for device');
+            }
+            if (pinCode === undefined || pinCode === null) {
+                await entity.command(
+                    'closuresDoorLock',
+                    'clearPinCode',
+                    {
+                        'userid': userId,
+                    },
+                    getOptions(meta.mapped),
+                );
+            } else {
+                if (isNaN(pinCode)) {
+                    throw new Error('pinCode must be a number or pinCode');
+                }
+                await entity.command(
+                    'closuresDoorLock',
+                    'setPinCode',
+                    {
+                        'userid': userId,
+                        'userstatus': 1,
+                        'usertype': 0,
+                        'pincodevalue': pinCode.toString(),
+                    },
+                    getOptions(meta.mapped),
+                );
+            }
+            return {readAfterWriteTime: 200};
+        },
+        convertGet: async (entity, key, meta) => {
+            const userId = meta && meta.message && meta.message.pin_code ? meta.message.pin_code.userId : undefined;
+            if ( isNaN(userId) ) {
+                throw new Error('userId must be numbers');
+            }
+            if (converters.pincode_lock.userIdInRange(meta, userId)) {
+                throw new Error('userId must be in range for device');
+            }
+            await converters.pincode_lock.getPinCode(entity, userId, meta);
+        },
+        getPinCode: async (entity, userId, meta) => {
+            await entity.command(
+                'closuresDoorLock',
+                'getPinCode',
+                {
+                    'userid': userId,
+                },
+                getOptions(meta.mapped),
+            );
+        },
+        onReconnect: async (entity, meta) => {
+            if (converters.pincode_lock.pinCodeCountSet(meta)) {
+                const pinCodeCount = meta.mapped.meta.pinCodeCount;
+                for (let i = 0; i < pinCodeCount; i++) {
+                    await converters.pincode_lock.getPinCode(entity, i, meta);
+                }
+            }
+        },
+        pinCodeCountSet: (meta) => meta &&
```

Comment:
When using this converter, I think we should force the definition to have a pinCodeCount defined, this will avoid extra checks and without it we are not able to retrieve all pind codes anyway.

A cleaner solution would be to issue some command to retrieve this number from the device (but I could find such command to do this?)

Reply:
Fair enough - it is restricted to pincode_lock. I agree - I could not find a way to retrieve the limit.